File: api/contracts/installation_settings.py
"""Self-hosted administrator settings. Credentials remain write-only."""
provider_names = {"type": "string", "enum": ["github", "google", "gitlab", "oidc"]}
revision = {"type": "integer", "minimum": 0}
source = {"type": "string", "enum": ["operator", "settings"]}
port = {"type": "integer", "minimum": 1, "maximum": 65535}
security = {"type": "string", "enum": ["starttls", "tls"]}

schemas["InstallationLoginProvider"] = obj({
    "provider": provider_names, "enabled": B,
    "client_id": {**S, "maxLength": 512}, "secret_configured": B,
    "issuer_url": {**S, "maxLength": 2048}, "revision": revision,
    "callback_url": S, "source": source, "encryption_ready": B,
}, ["provider", "enabled", "client_id", "secret_configured", "issuer_url", "revision", "callback_url", "encryption_ready"])
schemas["InstallationLoginProviderInput"] = obj({
    "enabled": B, "client_id": {**S, "maxLength": 512},
    "client_secret": {**S, "writeOnly": True, "maxLength": 8192},
    "clear_secret": B, "issuer_url": {**S, "maxLength": 2048}, "expected_revision": revision,
}, ["enabled", "client_id", "issuer_url", "expected_revision"])

smtp_fields = {
    "enabled": B, "host": {**S, "maxLength": 253}, "port": port, "security": security,
    "username": {**S, "maxLength": 256, "description": "At most 256 UTF-8 bytes, without control characters."},
    "from_email": {**S, "maxLength": 254, "description": "At most 254 UTF-8 bytes."},
}
schemas["InstallationSMTP"] = obj({
    "revision": revision, "source": source, **smtp_fields,
    "password_set": B, "encryption_ready": B,
}, ["revision", "source", *smtp_fields, "password_set", "encryption_ready"])
schemas["InstallationSMTPInput"] = obj({
    "expected_revision": revision, **smtp_fields,
    "password": {**S, "writeOnly": True, "maxLength": 4096, "description": "At most 4096 UTF-8 bytes, without NUL or line breaks. Omit to keep the stored password."},
    "clear_password": B,
}, ["expected_revision", *smtp_fields])
schemas["InstallationSMTPTestInput"] = obj({"expected_revision": revision}, ["expected_revision"])
schemas["InstallationSMTPTestResult"] = obj({"sent": B, "recipient": S}, ["sent", "recipient"])
for name in ["InstallationLoginProvider", "InstallationLoginProviderInput", "InstallationSMTP", "InstallationSMTPInput", "InstallationSMTPTestInput", "InstallationSMTPTestResult"]:
    schemas[name]["additionalProperties"] = False

for path, method, operation, response, request in [
    ("/installation/login-providers/{provider}", "get", "getInstallationLoginProvider", "InstallationLoginProvider", None),
    ("/installation/login-providers/{provider}", "put", "putInstallationLoginProvider", "InstallationLoginProvider", "InstallationLoginProviderInput"),
    ("/installation/smtp", "get", "getInstallationSMTP", "InstallationSMTP", None),
    ("/installation/smtp", "put", "putInstallationSMTP", "InstallationSMTP", "InstallationSMTPInput"),
    ("/installation/smtp/test", "post", "testInstallationSMTP", "InstallationSMTPTestResult", "InstallationSMTPTestInput"),
]:
    route(path, method, operation, ref(response), ref(request) if request else None)
    item = paths[path][method]
    item["responses"]["403"] = {"description": "Self-hosted installation administrator access required"}
    if method != "get":
        item["responses"]["409"] = {"description": "Configuration revision changed"}
    if "{provider}" in path:
        item["parameters"].append({"name": "provider", "in": "path", "required": True, "schema": provider_names})

File: web/src/lib/installation-settings.ts
import { useEditionFeatures } from './dashboard-edition'
import { useQuery } from '@tanstack/react-query'
import type { components } from './api.generated'
import { client, unwrap } from './client'
import { useScope } from './scope'

export const loginProviderNames = {
  github: 'GitHub',
  google: 'Google',
  gitlab: 'GitLab',
  oidc: 'Enterprise SSO',
} as const
export type LoginProvider = keyof typeof loginProviderNames
export const loginProviders = Object.keys(loginProviderNames) as LoginProvider[]
export type LoginProviderSettings = components['schemas']['InstallationLoginProvider']
export type SMTPSettings = components['schemas']['InstallationSMTP']
export type SecretAction = 'keep' | 'replace' | 'clear'

export function useAuthStatus() {
  return useQuery({
    queryKey: ['auth-status'],
    queryFn: ({ signal }) => unwrap(client.GET('/auth/status', { signal })),
    staleTime: 30000,
    retry: false,
  })
}

export function useInstallationAccess() {
  const scope = useScope()
  const status = useAuthStatus()
  const features = useEditionFeatures()
  return {
    status,
    admin: scope.identity.admin,
    allowed:
      scope.identity.admin && (status.data?.deployment_mode === 'self-hosted' || features.operator),
  }
}

export function loginProviderFeature(provider: LoginProvider) {
  return provider === 'oidc' ? 'enterprise_sso' : 'oauth_login'
}

export function prepareLoginProvider(
  settings: LoginProviderSettings,
  action: SecretAction,
  secret: string,
): components['schemas']['InstallationLoginProviderInput'] {
  if (action === 'replace' && !secret) throw new Error('Enter the new client secret.')
  if (
    settings.enabled &&
    (!settings.client_id.trim() ||
      action === 'clear' ||
      (action === 'keep' && !settings.secret_configured))
  )
    throw new Error('An enabled provider needs a client ID and client secret.')
  if (settings.provider === 'oidc' && (settings.enabled || settings.issuer_url)) {
    let issuer: URL
    try {
      issuer = new URL(settings.issuer_url)
    } catch {
      throw new Error('Enter the HTTPS issuer URL from your identity provider.')
    }
    if (
      issuer.protocol !== 'https:' ||
      issuer.username ||
      issuer.password ||
      issuer.search ||
      issuer.hash
    )
      throw new Error('Use an HTTPS issuer URL without credentials, query or fragment.')
  }
  return {
    enabled: settings.enabled,
    client_id: settings.client_id.trim(),
    issuer_url: settings.provider === 'oidc' ? settings.issuer_url.trim() : '',
    expected_revision: settings.revision,
    ...(action === 'replace' ? { client_secret: secret } : {}),
    ...(action === 'clear' ? { clear_secret: true } : {}),
  }
}

export function prepareSMTP(
  settings: SMTPSettings,
  action: SecretAction,
  password: string,
): components['schemas']['InstallationSMTPInput'] {
  if (action === 'replace' && !password) throw new Error('Enter the new SMTP password.')
  const bytes = (value: string) => new TextEncoder().encode(value).length
  if (bytes(settings.host.trim()) > 253) throw new Error('Use an SMTP host of at most 253 bytes.')
  if (
    bytes(settings.username.trim()) > 256 ||
    /[\u0000-\u001f\u007f-\u009f]/.test(settings.username)
  )
    throw new Error('Use an SMTP username of at most 256 bytes without control characters.')
  if (bytes(settings.from_email.trim()) > 254)
    throw new Error('Use a sender email of at most 254 bytes.')
  if (action === 'replace' && (bytes(password) > 4096 || /[\0\r\n]/.test(password)))
    throw new Error('Use an SMTP password of at most 4096 bytes without NUL or line breaks.')

  if (!Number.isInteger(settings.port) || settings.port < 1 || settings.port > 65535)
    throw new Error('Use a port from 1 to 65535.')
  if (settings.enabled && (!settings.host.trim() || !settings.from_email.trim()))
    throw new Error('Enabled email delivery needs an SMTP host and sender address.')
  if (
    settings.enabled &&
    settings.username.trim() &&
    (action === 'clear' || (action === 'keep' && !settings.password_set))
  )
    throw new Error('Enter a password for SMTP authentication, or leave the username empty.')
  return {
    expected_revision: settings.revision,
    enabled: settings.enabled,
    host: settings.host.trim(),
    port: settings.port,
    security: settings.security,
    username: settings.username.trim(),
    from_email: settings.from_email.trim(),
    ...(action === 'replace' ? { password } : {}),
    ...(action === 'clear' ? { clear_password: true } : {}),
  }
}
